flowey/flowey_core/src/github_context.rs

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+//! Core types and traits used to read GitHub context variables.
+
+use crate::node::{user_facing::GhContextVar, ClaimVar, NodeCtx, ReadVar, StepCtx};
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use std::collections::BTreeMap;
+
+#[derive(Serialize, Deserialize)]
+pub struct Head {
+    #[serde(rename = "ref")]
+    pub head_ref: String,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct GhContextVarReaderEventPullRequest {
+    pub head: Head,
+}
+
+pub enum Root {}
+pub enum Event {}
+
+#[derive(Clone)]
+pub struct GhVarState {
+    pub raw_name: String,
+    pub backing_var: String,
+    pub is_secret: bool,
+    pub is_object: bool,
+}
+
+pub struct GhContextVarReader<'a, S> {
+    pub ctx: NodeCtx<'a>,
+    pub _state: std::marker::PhantomData<S>,
+}
+
+impl<S> GhContextVarReader<'_, S> {
+    fn read_var<T: Serialize + DeserializeOwned>(
+        &self,
+        var_name: String,
+        is_secret: bool,
+        is_object: bool,
+    ) -> ReadVar<T> {
+        let (var, write_var) = self.ctx.new_maybe_secret_var(is_secret, "");
+        let write_var = write_var.claim(&mut StepCtx {
+            backend: self.ctx.backend.clone(),
+        });
+        let var_state = GhVarState {
+            raw_name: var_name.clone(),
+            backing_var: write_var.backing_var,
+            is_secret: write_var.is_secret,
+            is_object,
+        };
+        let gh_to_rust = vec![var_state];
+
+        self.ctx.backend.borrow_mut().on_emit_gh_step(
+            &format!("🌼 read {}", var_name),
+            "",
+            BTreeMap::new(),
+            None,
+            BTreeMap::new(),
+            BTreeMap::new(),
+            gh_to_rust,
+            Vec::new(),
+        );
+        var
+    }
+}
+
+impl<'a> GhContextVarReader<'a, Root> {
+    pub fn global(&self, gh_var: GhContextVar) -> ReadVar<String> {
```

Comment:
now that you have this `read_var` helper, I _really_ think that it's not that hard to swap the current `const`ants with standalone helpers - especially since you've already needed to touch those files as part of this `global` method refactoring.

the only thing you'll need to tweak is the user-defined secret var handling, but even that should be pretty straightforward, if you simply swap the current type that flow uses (i.e: the `GhContextVar` type we want to go away), with a far more limited newtype around a `String` called `GhUserSecretVar`, which is what gets created at the Pipeline level, and plumbed down to a new `pub fn secret()` method that hangs off this `GhContextVarReader`.